cpu/cc3200/periph/gpio.c

```diff
@@ -0,0 +1,393 @@
+ */
+unsigned long gpio_pin_to_port(uint8_t port)
+{
+    return ports[port];
+}
+
+/**
+ * @brief returns the masked value for a external pin
+ *
+ * @param pin external pin
+ * @param val value to be masked
+ * @return masked value for a given pin with value val
+ */
+unsigned char _gpio_pin_value_mask(uint8_t pin, unsigned char val)
+{
+    return val << (pin % 8);
+}
+
+/**
+ * @brief Access GPIO low-level device
+ *
+ * @param[in] pin   gpio pin
+ *
+ * @return          pointer to gpio low level device address
+ */
+static inline cc3200_gpio_t *gpio(gpio_t pin)
+{
+    return (cc3200_gpio_t *)(gpio_pin_to_port(gpio_port_num(pin)));
+}
+
+void gpio_init_af(gpio_t dev, uint32_t strength, uint32_t type)
+{
+    /* does not support analog pin types, but not a problem for GPIO */
+    uint8_t pin      = gpio_pin_num(dev);
+    uint8_t gpio_pin = pin_to_gpio_num[pin];
+
+    /* copied from TIs PinConfigSet. The register is not documented so for */
+    /* now only replecate behaviour. */
+
+    /* enable input */
+    HWREG(OCP_SHARED_BASE + OCP_SHARED_O_GPIO_PAD_CMN_CONFIG) &=
+            ~((0x80 << gpio_pin) & (0x1E << 8));
+
+    /* write config to hardware register referred by TI as PAD */
+    PAD_CONFIG_REG(gpio_pin) =
+            ((PAD_CONFIG_REG(gpio_pin) & ~(PAD_STRENGTH_MASK | PAD_TYPE_MASK)) |
+             (strength | type));
+}
+
+void gpio_pin_mode_set(gpio_t dev, uint32_t mode)
+{
+    /* does not support analog pin types, but not a problem for GPIO */
+    uint8_t pin      = gpio_pin_num(dev);
+    uint8_t gpio_pin = pin_to_gpio_num[pin];
+
+    /* set mode */
+    PAD_CONFIG_REG(gpio_pin) =
+            (((PAD_CONFIG_REG(gpio_pin) & ~PAD_MODE_MASK) | mode) & ~(3 << 10));
+}
+
+int gpio_init(gpio_t dev, gpio_mode_t mode)
+{
+    /* get pin mask */
+    uint8_t ipin = gpio_pin_mask(dev);
+
+    /* make sure pin is in the default state (this is quicker then reading */
+    /* first) */
+    gpio_init_af(dev, PIN_STRENGTH_2MA, PIN_TYPE_STD);
+    gpio_pin_mode_set(dev, PIN_MODE_GPIO);
+
+    /* set gpio direction IN/OUT */
+    if (mode & GPIO_DIR_MASK) {
+        /* out */
+        gpio(dev)->dir |= ipin;
+    } else {
+        /* in */
+        gpio(dev)->dir &= ipin;
+    }
+
+    /* set strength and drain mode */
+    gpio_init_af(dev, PIN_STRENGTH_2MA, mode & PAD_TYPE_MASK);
+
+    return 0;
+}
+
+#ifdef MODULE_PERIPH_GPIO_IRQ
+
+void isr_gpio_a0(void)
+{
+    handle_isr((cc3200_gpio_t *)GPIOA0_BASE);
+}
+
+void isr_gpio_a1(void)
+{
+    handle_isr((cc3200_gpio_t *)GPIOA1_BASE);
+}
+
+void isr_gpio_a2(void)
+{
+    handle_isr((cc3200_gpio_t *)GPIOA2_BASE);
+}
+
+void isr_gpio_a3(void)
+{
+    handle_isr((cc3200_gpio_t *)GPIOA3_BASE);
+}
+
+/**
+ * @brief isr interrupt handler
+ *
+ * @param portAddr base address of the GPIO PORT
+ */
+void handle_isr(cc3200_gpio_t *u)
+{
+    uint32_t state = u->mis;
+
+    /* clear interrupt */
+    u->icr = state;
+    for (int i = 0; i < 8; i++) {
+        if (state & (1 << i)) {
```

Comment:
would this be more efficient or idiomatic? I personally find it more difficult to understand but that can be due to my limited c experience.

Reply:
It would be a bit more efficient, since if only one bit is set it will only loop once. `clz` is *Count Leading Zeros*, many CPUs have a [hardware instruction](http://www.keil.com/support/man/docs/armasm/armasm_dom1361289868426.htm) for it.

Reply:
If there is interest / necessity in testing the functionality, we have a few of this boards available and I could send one (or try making on available remotely)

Reply:
That would be nice :)

A new platform needs someone to maintain it or else it will fall victim to bit rot and will eventually be removed.

If some more people have the hardware to test, this can be somewhat mitigated. One of the fun parts of being a RIOT maintainer is having a good collection of hardware to play with.

But I will still expect you to care for this platform and not drop the code and leave once it's merged 😉

Reply:
I will go ahead and update the loop then 👍 

As it looks right now I will be sticking to this platform for some time. The big plan is to port 802.11s to the platform (it is tricky but I know that it can be done). Since it is more of a fun/hobby project it will definitely take some time :) . At some point I will most likely switch to the CC3220 due to its better Wifi raw socket support. With that I would probably generalise the port for the whole CC32xx platform.
So no worries I am not going to abandon this board too quickly. :)

Reply:
That'd be great! Right now I think the Ti platforms are a bit short in love they receive. There is potentially a lot of code duplication between cc2538,  cc26x0, cc3200 and possibly lm4f120 too as hardware vendors tend to have a set of peripheral IP that they just reuse for their entire line-up of SoCs.

Reply:
> If some more people have the hardware to test, this can be somewhat mitigated. One of the fun parts of being a RIOT maintainer is having a good collection of hardware to play with.

Fair enough. I meant more for the initial testing. The greater plan is to include it into an iot-lab.info-like testbed which supports multiple boards with multiple networking technolgogies. That's one reason for the CC3200/CC3220, as they are pretty much the only ones, where the specs looked as if WiFi support seems feasible (which -as we learned - was not entirely true :-) )